packages/techdocs-common/src/stages/publish/awsS3.ts

```diff
@@ -0,0 +1,224 @@
+    private readonly bucketName: string,
+    private readonly logger: Logger,
+  ) {
+    this.storageClient = storageClient;
+    this.bucketName = bucketName;
+    this.logger = logger;
+  }
+
+  /**
+   * Upload all the files from the generated `directory` to the S3 bucket.
+   * Directory structure used in the bucket is - entityNamespace/entityKind/entityName/index.html
+   */
+  publish({ entity, directory }: PublishRequest): Promise<void> {
+    return new Promise(async (resolve, reject) => {
+      // Note: S3 manages creation of parent directories if they do not exist.
+      // So collecting path of only the files is good enough.
+      const allFilesToUpload = await getFileTreeRecursively(directory);
+
+      const uploadPromises: Array<Promise<PutObjectCommandOutput>> = [];
+      allFilesToUpload.forEach(filePath => {
+        // Remove the absolute path prefix of the source directory
+        // Path of all files to upload, relative to the root of the source directory
+        // e.g. ['index.html', 'sub-page/index.html', 'assets/images/favicon.png']
+        const relativeFilePath = filePath.replace(`${directory}/`, '');
+        const entityRootDir = `${entity.metadata.namespace}/${entity.kind}/${entity.metadata.name}`;
+        const destination = `${entityRootDir}/${relativeFilePath}`; // S3 Bucket file relative path
+
+        fs.readFile(filePath, (err, data) => {
+          if (err) {
+            reject(err);
+          }
+
+          const params = {
+            Bucket: this.bucketName,
+            Key: destination,
+            Body: data,
+          };
+
+          // TODO: Upload in chunks of ~10 files instead of all files at once.
+          uploadPromises.push(this.storageClient.putObject(params));
+        });
+      });
+      Promise.all(uploadPromises)
+        .then(() => {
+          this.logger.info(
+            `Successfully uploaded all the generated files for Entity ${entity.metadata.name}. Total number of files: ${allFilesToUpload.length}`,
+          );
+          resolve(undefined);
+        })
+        .catch((err: Error) => {
+          const errorMessage = `Unable to upload file(s) to AWS S3. Error ${err.message}`;
+          this.logger.error(errorMessage);
+          reject(errorMessage);
+        });
+    });
+  }
+
+  fetchTechDocsMetadata(entityName: EntityName): Promise<string> {
+    return new Promise((resolve, reject) => {
+      const entityRootDir = `${entityName.namespace}/${entityName.kind}/${entityName.name}`;
+
+      this.storageClient
+        .getObject({
+          Bucket: this.bucketName,
+          Key: `${entityRootDir}/techdocs_metadata.json`,
+        })
+        .then(file => {
+          const techdocsMetadataJson = file?.Body?.toString();
+
+          if (!techdocsMetadataJson) {
+            throw new Error(
+              `Unable to parse the techdocs metadata file ${entityRootDir}/techdocs_metadata.json.`,
+            );
+          }
```

Comment:
Do we have another way to get the `Readable | ReadableStream | Blob` in a better way ?